### PR TITLE
Add the smooth formulation of augmented Lagrangian.

### PR DESCRIPTION
**Pull request by hongkai-dai**
The slack variable s becomes a decision variable in the augmented Lagrangian. This is the same formulation used in LANCELOT.

Towards #16442

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@mpetersen94 for feature review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845#-:-MyoVTOt1Ifydc7cWdKm:b3ej2ph)* status: LGTM missing from assignee mpetersen94, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @mpetersen94)

**Review by hongkai-dai — COMMENTED** [review]
-@mpetersen94 for feature review, +@aykut-tri for feature review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845#-:-MzW5Sn-07mEiiPXAcch:bvh8g1x)* status: LGTM missing from assignee aykut-tri, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @aykut-tri)

**Comment by aykut-tri** [comment]
*[solvers/augmented_lagrangian.h, line 239 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqRAgs1j0hKl_P5Oc2:-MzqRAgs1j0hKl_P5Oc3:blqti5x) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.h#L239)):*
> ```objc
>   }
> 
>   /** @return all the lower bounds of x. */
> ```

Should the first letter of 'all' be capitalized as consistent with the other return descriptions?

**Comment by aykut-tri** [comment]
*[solvers/augmented_lagrangian.h, line 242 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqRRZpCd2lXiTo2C92:-MzqRRZpCd2lXiTo2C93:b-ya81ng) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.h#L242)):*
> ```objc
>   [[nodiscard]] const Eigen::VectorXd& x_lo() const { return x_lo_; }
> 
>   /** @return all the upper bounds of x. */
> ```

Same as above

**Comment by aykut-tri** [comment]
*[solvers/augmented_lagrangian.cc, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqS7eO2J1UEdAEvJXY:-MzqS7eO2J1UEdAEvJXZ:b54rq8w) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.cc#L30)):*
> ```cpp
> 
> // Compute the augmented lagrangian for the equality constraint c(x) = 0
> // The augmented lagrangian is −λc+μ/2*c² where c is lhs.
> ```

Consider capitalizing the first letter of Lagrangian for consistency



_Code quote:_
```cpp
/
```

**Comment by aykut-tri** [comment]
*[solvers/augmented_lagrangian.cc, line 162 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqU-II3FJ40B1IogCw:-MzqU-II3FJ40B1IogCx:b-bhhn5f) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.cc#L162)):*
> ```cpp
>               "constraint lower bound is {}, upper bound is {}", lb, ub));
>         }
>         if (lb == ub) {
> ```

I understand that this is probably out of the scope of this CR but just curious whether an approximate equality check would be more proper here? (Feel free to ignore and resolve this comment)

**Comment by aykut-tri** [comment]
*[solvers/augmented_lagrangian.cc, line 164 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqVHpE9mxOynbCa4Pl:-MzqVHpE9mxOynbCa4Pm:bls6jo5) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.cc#L164)):*
> ```cpp
>         if (lb == ub) {
>           // We have one Lagrangian multiplier for the equality constraint. Add
>           // −λ h(x) + μ/2*h(x)² to the augmented Lagrangian.
> ```

Missing multiplication sign in between



_Code quote:_
```cpp
−λ h(x)
```

**Comment by aykut-tri** [comment]
*[solvers/test/augmented_lagrangian_test.cc, line 212 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-Mzqc3h_-VeAXqTVb33m:-Mzqc3h_-VeAXqTVb33n:b6rw27d) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/test/augmented_lagrangian_test.cc#L212)):*
> ```cpp
>       EXPECT_EQ(dut.is_equality(), std::vector<bool>({true, true, true, true}));
>       al_val =
>           dut.Eval<T>(x_val, lambda_val, mu_val, &constraint_residue, &cost);
> ```

Shall we eliminate the repetition here by only defining `dut` based on the flag?



_Code quote:_
```cpp
      EXPECT_EQ(dut.lagrangian_size(), 4);
      EXPECT_EQ(dut.is_equality(), std::vector<bool>({true, true, true, true}));
      al_val =
          dut.Eval<T>(x_val, lambda_val, mu_val, &constraint_residue, &cost);
```

**Comment by aykut-tri** [comment]
*[bindings/pydrake/solvers/test/augmented_lagrangian_test.py, line 91 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqeniqAJK4fWMEEWdI:-MzqenirCOyBErxDe4yc:bptbi4a) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/bindings/pydrake/solvers/test/augmented_lagrangian_test.py#L91)):*
> ```Python
>         self.assertEqual(
>             al.AugmentedLagrangianSmooth(prog=self.prog,
>                                          include_x_bounds=True).is_equality(),
> ```

Is it worth writing this in the above form (for the non-smooth version) to save a line?



_Code quote:_
```Python
            al.AugmentedLagrangianSmooth(prog=self.prog,
                                         include_x_bounds=True).is_equality(),
            [False])
```

**Review by aykut-tri — COMMENTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845)* status: 7 unresolved discussions, LGTM missing from assignee aykut-tri, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845)* status: 7 unresolved discussions, LGTM missing from assignee aykut-tri, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @aykut-tri)

___
*[solvers/augmented_lagrangian.h, line 239 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqRAgs1j0hKl_P5Oc2:-MzsBlrsCMQ8cJgoNJFZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.h#L239)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

Should the first letter of 'all' be capitalized as consistent with the other return descriptions?
</blockquote></details>

Done.
___
*[solvers/augmented_lagrangian.h, line 242 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqRRZpCd2lXiTo2C92:-MzsBmK871gLdf4P-Nsx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.h#L242)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.
___
*[solvers/augmented_lagrangian.cc, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqS7eO2J1UEdAEvJXY:-MzsBt-V2c7P8nCiHwFV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.cc#L30)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

Consider capitalizing the first letter of Lagrangian for consistency
</blockquote></details>

Done.
___
*[solvers/augmented_lagrangian.cc, line 162 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqU-II3FJ40B1IogCw:-MzsBvFpAQtyWM409M23:bizazpp) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.cc#L162)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

I understand that this is probably out of the scope of this CR but just curious whether an approximate equality check would be more proper here? (Feel free to ignore and resolve this comment)
</blockquote></details>

Could you elaborate why doing approximate check instead of exact check?

The bounds of the constraint comes from the user when creating the constraints. I think if the user really think the constraints is equality, he/she should have set the bounds to be exactly the same?
___
*[solvers/augmented_lagrangian.cc, line 164 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqVHpE9mxOynbCa4Pl:-MzsCFYzATMLpOa19szW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.cc#L164)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

Missing multiplication sign in between
</blockquote></details>

Done.
___
*[solvers/test/augmented_lagrangian_test.cc, line 212 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-Mzqc3h_-VeAXqTVb33m:-MzsCUmg0t2dg-maWfes:b-uqsjnj) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/test/augmented_lagrangian_test.cc#L212)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

Shall we eliminate the repetition here by only defining `dut` based on the flag?
</blockquote></details>

The problem is that these two classes have different signature in Eval function. For AugmentedLagrangianSmooth::Eval, it has the extra input `s`, while for the non-smooth version, s is not an input but computed within Eval from `x`. This difference in the Eval function API is the reason why I didn't combine these two classes into one class.

I added documentation in the header file to make it more clear.
___
*[bindings/pydrake/solvers/test/augmented_lagrangian_test.py, line 91 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqeniqAJK4fWMEEWdI:-MzsDYZQ5PWqYpqYMz96:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/bindings/pydrake/solvers/test/augmented_lagrangian_test.py#L91)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

Is it worth writing this in the above form (for the non-smooth version) to save a line?
</blockquote></details>

Done.

**Review by aykut-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845)* status: 1 unresolved discussion, LGTM missing from assignee aykut-tri, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

**Comment by aykut-tri** [comment]
*[solvers/augmented_lagrangian.cc, line 162 at r1](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzqU-II3FJ40B1IogCw:-MzuOEW2BKZAmNqwyKFV:bpwwsye) ([raw file](https://github.com/robotlocomotion/drake/blob/32551f3bbbf07e5cd39e277c6e9ea226c8da057c/solvers/augmented_lagrangian.cc#L162)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I see.

I think the issue is that here in AugmentedLagrangian class we do not know that solver tolerance. First each solver has a different tolerance (SNOPT/IPOPT/NLOPT) and here we don't know which solver is used. Second the user can adjust the solver's default tolerance to his specified value and we won't know the adjusted tolerance in this class. Hence I would like to use the exact equality here. Does that make sense?
</blockquote></details>

Yeap, absolutely.

**Comment by aykut-tri** [comment]
:lgtm:

**Review by hongkai-dai — COMMENTED** [review]
+@rpoyner-tri for platform review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845#-:-MzuOZbP4wB-kqnJGozG:b8asji7)* status: LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @rpoyner-tri)

**Review by rpoyner-tri — COMMENTED** [review]
Checkpoint. Will resume after meetings.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845#-:-MzuW8ro4EvhOwAsOHMQ:b-159er6)* status: 5 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @hongkai-dai)

___
*[solvers/augmented_lagrangian.h, line 183 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuRsbO5lKCxetW2S7V:-MzuRsbO5lKCxetW2S7W:bi5cjj2) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L183)):*
> ```objc
>    * @param s The value of all slack variables s.
>    * @param lambda_val The estimated Lagrangian multipliers. The order of the
>    * Lagrangian multiplier is as this: We first call to evaluate all
> ```

nit this phrasing is awkward. Consider s/is as this:/is as follows./



_Code quote:_
```objc
is as this:
```
___
*[solvers/augmented_lagrangian.h, line 251 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuVoBj86tVJI5oViuv:-MzuVoBk318NxDyiQwEX:b-mijpn2) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L251)):*
> ```objc
>   bool include_x_bounds_;
>   int lagrangian_size_;
>   int s_size_;
> ```

This variable is currently uninitialized until Eval() is called. Clients could call s_size() before that time and receive nasal demons. Consider initializing it to 0 here or at least `int s_size_{};`.



_Code quote (from solvers/augmented_lagrangian.cc):_
```cpp
s_size_
```
___
*[solvers/augmented_lagrangian.h, line 252 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuTeFTCck0IiJUtq2D:-MzuTeFTCck0IiJUtq2E:bcl6f6u) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L252)):*
> ```objc
>   int lagrangian_size_;
>   int s_size_;
>   std::vector<bool> is_equality_;
> ```

nit You are aware that the memory layout of vector<bool> is packed bits, right? It may not matter in this case, but can be a source of trouble (no operations are atomic, size assumptions are probably wrong, slowness, etc.).
___
*[solvers/augmented_lagrangian.cc, line 21 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuUIX04Bi7-8zfHM0v:-MzuUIX04Bi7-8zfHM0w:b9344c3) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L21)):*
> ```cpp
> // use for μ is actually 1/μ in our formulation.
> template <typename T>
> T psi(const T& c, double lambda_val, double mu) {
> ```

minor, maybe out of scope: this function is not mentioned in the header, and yet is within a public namespace in the .cc file. The potential for link-time name collisions exists. Can this be in the anonymous namespace?
___
*[solvers/augmented_lagrangian.cc, line 32 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuUeu06gmX9rU_eAyc:-MzuUeu06gmX9rU_eAyd:b6p2443) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L32)):*
> ```cpp
> // The augmented Lagrangian is −λc+μ/2*c² where c is lhs.
> template <typename T>
> T al_for_equality(const T& lhs, double lambda_val, double mu) {
> ```

minor: ditto above. Can this be in the anonymous namespace?

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845)* status: 6 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @hongkai-dai)

___
*[solvers/augmented_lagrangian.cc, line 37 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuZGxwAWMRCGZRJVvG:-MzuZGxwAWMRCGZRJVvH:b-7tqdl0) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L37)):*
> ```cpp
> 
> namespace {
> void ParseProgram(const MathematicalProgram* prog, bool include_x_bounds,
> ```

nit: Considering that none of the primitive-type member variables are initialized in the class declaration, but some of them are initialized here, I find it somewhat hard to follow the logic. Would it be clearer to have ParseProgram() be a private method (with many fewer parameters)? Since this is an important initialization helper, the declaration could possibly include comments about what members are initialized.
___
*[solvers/augmented_lagrangian.cc, line 59 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-Mzu_6FQAy08djIjBhsu:-Mzu_6FQAy08djIjBhsv:bhyrcdq) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L59)):*
> ```cpp
>     }
>   }
>   AggregateBoundingBoxConstraints(*prog, x_lo, x_up);
> ```

nit: ditto above. Would this be clearer/simpler as a private method?

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845)* status: 3 unresolved discussions, LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @rpoyner-tri)

___
*[solvers/augmented_lagrangian.h, line 251 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuVoBj86tVJI5oViuv:-MzvAMhS5MWAFv_9_E0k:b-fuojd0) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L251)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

This variable is currently uninitialized until Eval() is called. Clients could call s_size() before that time and receive nasal demons. Consider initializing it to 0 here or at least `int s_size_{};`.
</blockquote></details>

Done. But this variable is initialized in the constructor of the class in https://github.com/hongkai-dai/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L262-L268
___
*[solvers/augmented_lagrangian.h, line 252 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuTeFTCck0IiJUtq2D:-MzvAoKvC-nyrUQh-AC3:b-xbsgtc) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L252)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

nit You are aware that the memory layout of vector<bool> is packed bits, right? It may not matter in this case, but can be a source of trouble (no operations are atomic, size assumptions are probably wrong, slowness, etc.).
</blockquote></details>

So should I use some other data types? I do have an array of bools and I don't know the size of that array before constructing this class.
___
*[solvers/augmented_lagrangian.cc, line 21 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuUIX04Bi7-8zfHM0v:-MzvBD1W4vYpHXPwH_Cd:b-yh16l6) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L21)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

minor, maybe out of scope: this function is not mentioned in the header, and yet is within a public namespace in the .cc file. The potential for link-time name collisions exists. Can this be in the anonymous namespace?
</blockquote></details>

Done. Good call.
___
*[solvers/augmented_lagrangian.cc, line 32 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuUeu06gmX9rU_eAyc:-MzvBFcNB2fVnsnrJodN:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L32)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

minor: ditto above. Can this be in the anonymous namespace?
</blockquote></details>

Done.
___
*[solvers/augmented_lagrangian.cc, line 37 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuZGxwAWMRCGZRJVvG:-MzvBM9M1yeknyFcHARD:b-9wqo4q) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L37)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

nit: Considering that none of the primitive-type member variables are initialized in the class declaration, but some of them are initialized here, I find it somewhat hard to follow the logic. Would it be clearer to have ParseProgram() be a private method (with many fewer parameters)? Since this is an important initialization helper, the declaration could possibly include comments about what members are initialized.
</blockquote></details>

I intend to call this ParseProgram in both classes (AugmentedLagrangianNonsmooth and AugmentedLagrangianSmooth). Should I add a private function `ParseProgram()` to both classes, and this `ParseProgram` function call another shared function?

I can also keep this current ParseProgram function unchanged, but add documentations, would that help?
___
*[solvers/augmented_lagrangian.cc, line 59 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-Mzu_6FQAy08djIjBhsu:-MzvBs_n1V-jwjHX_mza:b6sndks) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L59)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

nit: ditto above. Would this be clearer/simpler as a private method?
</blockquote></details>

Not sure if we need to put this as a private method, it is currently a public method in https://github.com/RobotLocomotion/drake/blob/520105b6940035cb4b50d93e2f3946115f6edba4/solvers/aggregate_costs_constraints.h#L69-L82

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm:

> Reviewed 2 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845#-:-MzvHRiZ3RRYyT-mf8gR:bnfp4nl)* status: missing label for release notes (waiting on @hongkai-dai)

___
*[solvers/augmented_lagrangian.h, line 251 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuVoBj86tVJI5oViuv:-MzvDv_9EscHOpNjF2Ah:b-9c3pz) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L251)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done. But this variable is initialized in the constructor of the class in https://github.com/hongkai-dai/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L262-L268
</blockquote></details>

Sorry, I misread the code the first time through. I tried to retract the comment, but apparently failed :/ .
___
*[solvers/augmented_lagrangian.h, line 252 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuTeFTCck0IiJUtq2D:-MzvE8MoC0rPo_qzeuGu:b-l6709c) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L252)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

So should I use some other data types? I do have an array of bools and I don't know the size of that array before constructing this class.
</blockquote></details>

I think it is OK in this case; I just know that decades of C++ community regret about `vector<bool>` makes me take notice when it gets used. Since you never try to (say) take a reference to a contained value, I think we are fine here.

For historical reference of concerns, this page is a good start: https://code-examples.net/en/q/10f8609
___
*[solvers/augmented_lagrangian.cc, line 37 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuZGxwAWMRCGZRJVvG:-MzvG6Lj1Rq54XQ2oMbe:bp9humq) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L37)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I intend to call this ParseProgram in both classes (AugmentedLagrangianNonsmooth and AugmentedLagrangianSmooth). Should I add a private function `ParseProgram()` to both classes, and this `ParseProgram` function call another shared function?

I can also keep this current ParseProgram function unchanged, but add documentations, would that help?
</blockquote></details>

If you intend to share it across both classes, then yes, having it as an anonymous free function is fine. I was missing that piece of context in this change set.
___
*[solvers/augmented_lagrangian.cc, line 59 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-Mzu_6FQAy08djIjBhsu:-MzvGiKB7k5Evh3KZma8:b-bvavg4) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.cc#L59)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Not sure if we need to put this as a private method, it is currently a public method in https://github.com/RobotLocomotion/drake/blob/520105b6940035cb4b50d93e2f3946115f6edba4/solvers/aggregate_costs_constraints.h#L69-L82
</blockquote></details>

Nevermind. Misread.

**Review by rpoyner-tri — COMMENTED** [review]
+(release notes: yes) 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845#-:-MzvLkiNENm8xsZAuTBM:b-c7f1bi)* status: :shipit: complete! all discussions resolved, LGTM from assignees aykut-tri,rpoyner-tri(platform) (waiting on @hongkai-dai)

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/16845)* status: :shipit: complete! all discussions resolved, LGTM from assignees aykut-tri,rpoyner-tri(platform)

___
*[solvers/augmented_lagrangian.h, line 252 at r2](https://reviewable.io/reviews/robotlocomotion/drake/16845#-MzuTeFTCck0IiJUtq2D:-MzvQU_7-gwUwqFrMjzO:blml9j) ([raw file](https://github.com/robotlocomotion/drake/blob/654145f68a95280eb44317404cec628843e43348/solvers/augmented_lagrangian.h#L252)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

I think it is OK in this case; I just know that decades of C++ community regret about `vector<bool>` makes me take notice when it gets used. Since you never try to (say) take a reference to a contained value, I think we are fine here.

For historical reference of concerns, this page is a good start: https://code-examples.net/en/q/10f8609
</blockquote></details>

Thanks @rpoyner-tri , I didn't realize there was the problem of std::vector<bool>. TIL.
